### PR TITLE
[query] Expose TableGen to Python

### PR DESCRIPTION
**Pull request by ehigham**
Add python classes analogous to those in Scala
Add printer/parsers to Scala and Python

### CODE REVIEW

File: hail/python/hail/table.py [hunk 1]
```diff
@@ -554,6 +555,44 @@ def parallelize(cls, rows, schema=None, key=None, n_partitions=None, *,
             table = table.key_by(*key)
         return table
 
+    @staticmethod
+    @typecheck(
+        contexts=expr_stream(expr_any),
```
Comment (tpoterba) [inline hail/python/hail/table.py:560]: prefer expr_array here. We try not to expose streams much at the Python level -- just wrap the contexts IR in ToStream when creating the TableGen IR

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala [hunk 2]
```diff
@@ -796,23 +797,52 @@ object LowerTableIR {
...
-              s"${t.getClass.getSimpleName}: partitioner contains ${partitioner.numPartitions} partitions, got ",
-              invoke("str", TString, ArrayLen(ref))
-            )
-            ToStream(
-              If(ArrayLen(ref) ceq partitioner.numPartitions, ref, Die(dieMsg, ref.typ, traceId))
-            )
+          contexts = lowerIR {
+            bindIR(ToArray(contexts)) { ref =>
+              bindIR(ArrayLen(ref)) { len =>
+                // If contexts were an array we could assert this at compile time...
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:809]: not true -- arrays aren't statically sized in the general case. We could write an analysis that tries to infer static length, but not worth it to eliminate this check.

**Review by tpoterba — DISMISSED** [review]
2 nits -- nice work!

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala (hunk 2 above)
Comment (ehigham) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:809]: But would those arrays not be fixed size generated from python?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:809]: you're right that if someone is passing in a partitioner, someone at some point probably is generating an array of a fixed size. But we don't know the runtime length of any particular IR of type TArray by default.

Comment (ehigham) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:809]: fair enough!

File: hail/python/hail/table.py (hunk 1 above)
Comment (ehigham) [reply hail/python/hail/table.py:560]: How about the body return type?

**Review by tpoterba — APPROVED** [review]

**Pull request merged**